### ISSUE TITLE
performance is poor on valid set using e-commerce dataset

### ISSUE THREAD

**Issue by brightmart** [alibaba/esim-response-selection#6]
<img width="1399" alt="image" src="https://user-images.githubusercontent.com/19634224/62587414-79698400-b8f4-11e9-8120-cc1bc595fbb0.png">

is there someone can help or give advice on how to make it work?

change hidden size to 100, so can use batch size 32.

hyper-parameter as below:
----------------------------------
CUDA_VISIBLE_DEVICES=6 python -u main.py \
--train_file=$DATA_DIR/train.txt \
--valid_file=$DATA_DIR/valid.txt \
--test_file=$DATA_DIR/test.txt \
--vocab_file=$DATA_DIR/vocab.txt \
--output_dir=result \
--embedding_file=../../data/embedding_w2v_d300.txt \
--maxlen_1=300 \
--maxlen_2=150 \
--hidden_size=100 \
--train_batch_size=32 \
--valid_batch_size=16 \
--test_batch_size=16 \
--fix_embedding=True \
--patience=1 \
> log.txt 2>&1 &

**Comment by brightmart** [comment]
@lukecq1231 @alibaba-oss @enningxie

**Comment by panzhufeng** [comment]
Got the same problem. Looking forward to the reply.

**Comment by lukecq1231** [comment]
@brightmart @vufg Because the Ubuntu and E-commerce dataset have different number of candidates per context in valid set. Ubuntu has 10 candidates per context (1 positive, 9 negative), but E-commerce has only 2 candidates per context (1 positive, 1 negative). Thus, you need change the metrics for valid set, i.e., change the predict_metrics function to predict_accuracy function. 

You could change code at: 
line 724-728,

            valid_metrics, valid_scores = predict_metrics(
                sess, cost_op, probability_op, valid)

            # select best model based on recall@1 of validation set
            valid_err = 1.0 - valid_metrics[3]
            history_errs.append(valid_err)

            tf.logging.info(
                "valid set: MAP %s MRR %s Precision@1 %s Recall@1 %s Recall@2 %s Recall@5 %s", *valid_metrics)

to:

            valid_acc, valid_cost = predict_accuracy(
                sess, cost_op, probability_op, valid)

            # select best model based on accuracy of validation set
            valid_err = 1.0 - valid_acc
            history_errs.append(valid_err)

            tf.logging.info("valid set: ACC %s Cost %s", valid_acc, valid_cost)

line 758-761 from:

        valid_metrics, valid_scores = predict_metrics(
            sess, cost_op, probability_op, valid)
        tf.logging.info(
            "valid set: MAP %s MRR %s Precision@1 %s Recall@1 %s Recall@2 %s Recall@5 %s", *valid_metrics)

to:

        valid_acc, valid_cost = predict_accuracy(
            sess, cost_op, probability_op, valid)

        tf.logging.info("valid set: ACC %s Cost %s", valid_acc, valid_cost)

**Comment by Saozy** [comment]
The test set of E-com has 1 positive sample and 9 negtive samples. i wonder... why the precison and recall are both about 0.5 after being trained for 10 epochs.

**Comment by lukecq1231** [comment]
@Saozy What is your hyper-parameters? I used following hyper-parameters and could get a good result.

INFO:tensorflow:valid set: ACC 0.9395 Cost 0.16921886469423772
INFO:tensorflow:seen samples 10000
INFO:tensorflow:test set: MAP 0.749915079365 MRR 0.749915079365 Precision@1 0.61 Recall@1 0.61 Recall@2 0.777 Recall@5 0.951


	DATA_DIR=../../data/E-commerce_dataset_concat

	CUDA_VISIBLE_DEVICES=0 python -u main.py \
	--train_file=$DATA_DIR/train.txt \
	--valid_file=$DATA_DIR/dev.txt \
	--test_file=$DATA_DIR/test.txt \
	--vocab_file=$DATA_DIR/vocab.txt \
	--output_dir=result \
	--embedding_file=../../data/embedding_w2v_d300_taobao.txt \
	--maxlen_1=300 \
	--maxlen_2=50 \
	--hidden_size=300 \
	--vocab_size=36136 \
	--train_batch_size=32 \
	--valid_batch_size=32 \
	--test_batch_size=32 \
	--fix_embedding=False \
	--learning_rate=4e-4 \
	--patience=1 \
	> log.txt 2>&1 &

**Comment by Saozy** [comment]
@lukecq1231 i used your hyper-p and got a result like yours, thx
